include/kernels/madwf_tensor.cuh

```diff
@@ -0,0 +1,205 @@
+                                                         const WilsonVector<real> &w)
+    {
+      complex<real> *p = reinterpret_cast<complex<real> *>(mp);
+
+#pragma unroll
+      for (int a = 0; a < spin_dim; a++) {
+#pragma unroll
+        for (int b = 0; b < spin_dim; b++) {
+          int cs = a * spin_dim + b;
+          complex<real> z = 0;
+#pragma unroll
+          for (int color = 0; color < color_dim; color++) { z += conj(conj(v(a, color)) * w(b, color)); }
+          // Perform a block reduction across the x direction
+
+          block_reduce_x(z);
+
+          if (target::thread_idx().x == 0) {
+            p[(m_index * spin_dim * spin_dim + cs) * target::grid_dim().x + target::block_idx().x] = z;
+          }
+        }
+      }
+    }
+
+    template <class storage_type, class matrix_type_, int block_dim_x_> struct Tensor5DArg : kernel_param<> {
+
+      using F = typename colorspinor_mapper<storage_type, 4, 3>::type;
+      using real = typename mapper<storage_type>::type;
+      using Vector = ColorSpinor<real, 3, 4>;
+      using matrix_type = matrix_type_;
+
+      static constexpr int block_dim_x = block_dim_x_;
+
+      const F out; // output vector field
+      const F in;  // input vector field
+
+      const int Ls_out; // length of 5th dimension of the out field
+      const int Ls_in;  // length of 5th dimension of the in field
+
+      const int volume_4d_cb;
+
+      matrix_type *reduce_buffer = nullptr;
+
+      const int nParity;
+
+      matrix_type *result_d;
+
+      int batch;
+
+      Tensor5DArg(const ColorSpinorField &out, const ColorSpinorField &in, int num_x_blocks, matrix_type *result_d) :
+        kernel_param(dim3(out.VolumeCB() / out.X(4), out.X(4), out.SiteSubset())),
+        out(out),
+        in(in),
+        Ls_out(out.X(4)),
+        Ls_in(in.X(4)),
+        volume_4d_cb(in.VolumeCB() / in.X(4)),
+        nParity(in.SiteSubset()),
+        result_d(result_d),
+        batch(num_x_blocks)
+      {
+
+        if (volume_4d_cb != static_cast<int>(out.VolumeCB() / Ls_out)) {
+          errorQuda("Input and Output fields should have the same 4d volume: %d neq %d.\n", volume_4d_cb,
+                    static_cast<int>(out.VolumeCB() / Ls_out));
+        }
+
+        if (in.Nspin() != 4) errorQuda("nSpin = %d not support", in.Nspin());
+        if (in.Ncolor() != 3) errorQuda("nColor = %d not support", in.Ncolor());
+        if (out.Nspin() != 4) errorQuda("nSpin = %d not support", out.Nspin());
+        if (out.Ncolor() != 3) errorQuda("nColor = %d not support", out.Ncolor());
+
+        if (!in.isNative() || !out.isNative())
+          errorQuda("Unsupported field order out=%d in=%d\n", out.FieldOrder(), in.FieldOrder());
+
+        size_t reduce_bytes = num_x_blocks * sizeof(matrix_type) * out.X(4) * in.X(4);
+        reduce_buffer = reinterpret_cast<matrix_type *>(device_malloc(reduce_bytes));
```

Comment:
We should keep the argument classes trivially copyable.  So best here to move the reduce_buffer allocation and free outside of the constructor / destructor.

Reply:
The need to have the buffer has been removed.